### PR TITLE
[UBS Address] User can choose districts on own 

### LINKED ISSUE #6010: [UBS Address] User can not choose districts on own
It is not possible to choose regions, cities, and districts on their own, all of that is hard coded.

Expected result:
User can choose all elements of locations on own

### PR DESCRIPTION
**Pull request by ospodaryk**
# GreenCityUBS PR
[[UBS Address] User can not choose districts on own #6010
](https://github.com/ita-social-projects/GreenCity/issues/6010)

## Summary Of Changes :fire:
Added new API service which get region and city name and return district name.

## Added
New function to get all districts for location

# PR Checklist Forms
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

File: core/src/main/java/greencity/controller/AddressController.java [hunk 1]
```diff
@@ -15,18 +16,11 @@
...
-import org.springframework.web.bind.annotation.DeleteMapping;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PatchMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.PutMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/AddressController.java]: please do not use `*`

File: core/src/main/java/greencity/controller/AddressController.java [hunk 2]
```diff
@@ -142,4 +136,19 @@ public ResponseEntity<AddressDto> makeAddressActual(
...
+
+    /**
+     * Controller to get all districts for a given region and city.
+     *
+     * @param region Name of the region.
+     * @param city   Name of the city.
+     * @return A List of LocationDtos containing a list of all districts for the
+     *         specified region and city.
+     */
+    @GetMapping("/get-all-districts")
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/AddressController.java:170]: please add `ApiOperation` and `ApiResponses`

File: core/src/test/java/greencity/controller/AddressControllerTest.java [hunk 3]
```diff
@@ -6,8 +6,10 @@
 import greencity.configuration.SecurityConfig;
 import greencity.converters.UserArgumentResolver;
 import greencity.dto.CreateAddressRequestDto;
+import greencity.dto.location.api.LocationDto;
 import greencity.dto.order.OrderAddressDtoRequest;
 import greencity.service.ubs.UBSClientService;
+import liquibase.pro.packaged.S;
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/AddressControllerTest.java]: where do you use it?

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+import lombok.Data;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.net.URI;
+import java.util.*;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: please do not use `*`

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    }
+
+    /**
+     * Retrieves all cities in a specific region.
+     *
+     * @param upperId The id of the upper region.
+     * @return A list of all cities in the specified region.
+     */
+    public List<LocationDto> getAllCitiesById(String upperId) {
+        return getLocationDataByUpperId(DEFAULT_PAGE_SIZE, 4, upperId);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: please do not use magic numbers

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    /**
+     * Retrieves a city in a specific region by its name.
+     *
+     * @param regionName The name of the region where the city is located.
+     * @param cityName   The name of the city.
+     * @return A LocationDto object containing the city's data.
+     */
+    public List<LocationDto> getAllDistrictsInCityByNames(String regionName, String cityName) {
+        if (cityName.equals("Київ") || cityName.equals("Kyiv")) {
+            return getAllDistrictsInCityByCityID("UA80000000000093317");
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: please move `Київ`, `Kyiv`, and  `UA80000000000093317` to the constants

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        }
+        List<LocationDto> allDistricts = null;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: Any reason you set `allDistricts = null`?
we can use 
`List<LocationDto> allDistricts = getAllDistrictsInCityByCityID(cityId);`

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+        List<LocationDto> cities = getCitiesByName(regionName, cityName);
+        LocationDto city = getCityInRegion(regionName, cities);
+        String cityId = city.getId();
+        allDistricts = getAllDistrictsInCityByCityID(cityId);
+        if (allDistricts.isEmpty()) {
+            String cityNameUK = city.getName().get("name");
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: can use use `NAME` here?

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+            city = getCitiesByName(regionName, cityNameUK).get(0);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: Can we have `indexoutofboud` exception here?

File: service/src/main/java/greencity/service/locations/LocationApiService.java [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+
+    /**
+     * Maps a result map into a LocationDto object.
+     *
+     * @param result a map containing the result data.
+     * @return a LocationDto object built from the result map.
+     */
+    private LocationDto mapToLocationDto(Map<String, Object> result) {
+        Map<String, String> nameMap = new HashMap<>();
+        nameMap.put("name", getValueFromMap(result, "name"));
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/locations/LocationApiService.java]: please use the constants you declared here https://github.com/ita-social-projects/GreenCityUBS/pull/1161/files#diff-dc3cf5a2a46824d3495f3c4da780b50ad4ddeebccfc29d59dd1c63d5d89a8152R21

File: service/src/test/java/greencity/service/ubs/UBSManagementEmployeeServiceImplTest.java [hunk 11]
```diff
@@ -23,17 +24,21 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
 import org.springframework.mock.web.MockMultipartFile;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
 
-import java.util.List;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ubs/UBSManagementEmployeeServiceImplTest.java]: please do not use `*`

**Review by sergoliarnik — APPROVED** [review]

**Review by RomaMocherniuk — APPROVED** [review]

**Review by ABbondar — APPROVED** [review]

**Review by Yurii-Midianyi — APPROVED** [review]

**Pull request merged**